Source/WebCore/dom/Element.cpp

```diff
@@ -2664,9 +2681,14 @@ void Element::removedFromAncestor(RemovalType removalType, ContainerNode& oldPar
 #endif
 
     if (UNLIKELY(hasRareData()) && !elementRareData()->effectiveLang().isNull()) {
+        ASSERT(!langFromAttribute().isNull());
         if (auto* parent = parentOrShadowHostElement(); langFromAttribute().isNull()
             && !(parent && UNLIKELY(parent->hasRareData()) && !parent->elementRareData()->effectiveLang().isNull()))
             elementRareData()->setEffectiveLang(nullAtom());
+    } else if (!langFromAttribute().isNull()) {
```

Comment:
I don't think we want this function call to happen whenever an element is removed.
The function is O(k) where k is the number of attributes on this element.
Instead, we should probably set a bit flag whenever we have an element with an explicit lang attribute.

Reply:
Also, we need to do this work only when removalType.disconnectedFromDocument is true.

Reply:
A bit flag is what I wanted to use, to avoid searching for the lang attribute, but I could not see a free flag on Node/Element.

Reply:
@heycam : There are more bits left in EventTarget.